src/realm/object_id.cpp

```diff
@@ -0,0 +1,101 @@
+/*************************************************************************
+ *
+ * Copyright 2019 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ **************************************************************************/
+
+#include <realm/object_id.hpp>
+#include <realm/util/assert.hpp>
+#include <chrono>
+#include <atomic>
+#include <random>
+
+using namespace std::chrono;
+
+static std::mt19937 generator(std::mt19937::result_type(system_clock::now().time_since_epoch().count()));
+
+static std::atomic<std::mt19937::result_type> seq(generator());
+static const char hex_digits[] = "0123456789abcdef";
+static const char null_id[12] = {0};
+
+namespace realm {
+
+ObjectId::ObjectId()
+{
+    memset(m_bytes, 0, sizeof(m_bytes));
+}
+
+ObjectId::ObjectId(const char* init)
+{
+    char buf[3];
+    REALM_ASSERT(strlen(init) == 24);
+
+    buf[2] = '\0';
+
+    size_t j = 0;
+    for (size_t i = 0; i < sizeof(m_bytes); i++) {
+        buf[0] = init[j++];
+        buf[1] = init[j++];
+        m_bytes[i] = char(strtol(buf, nullptr, 16));
+    }
+}
+
+ObjectId::ObjectId(Timestamp d, int machine_id, int process_id)
+{
+    auto sec = uint32_t(d.get_seconds());
+    auto as_bytes = type_punning<Byte4>(sec);
+    // Store in big endian so that memcmp can be used for comparison
```

Comment:
Doesn't this assume that everybody is running on little endian? I don't think we make that assumption anywhere else. I think an endian agnostic method could be `m_bytes[0] = sec >> 24; m_bytest[1] = (sec >> 16) & 0xFF; ...`

Reply:
As a minimum we should compile time assert that we are.

Reply:
I think we should continue to support both endianness. After all, REAL machines are big endian. Right?

Reply:
This code is just copied from realm-sync. I don't think we ever supported big endian machines with respect to this.

Reply:
Did as @ironage suggested.